parking_lot.py

```diff
@@ -0,0 +1,172 @@
+from datetime import datetime
+from typing import Optional, Union
+from enum import Enum
+from pydantic import BaseModel, Field, ValidationError, validator
+from fastapi import FastAPI, Request, status
+from starlette.responses import JSONResponse
+import math
+import dateutil.parser
+
+app = FastAPI()
+
+PARKED_CARS = {
+    'X773HY97': {'status': 'parked', 'tariff': 'hourly', 'location': 7, 'start': '2022-06-25T17:21:33.913233', 'end': None, 'fee': None},
+    'G737TT97': {'status': 'parked', 'tariff': 'daily', 'location': 14, 'start': '2022-06-25T17:23:24.428268', 'end': None, 'fee': None}
+}
+PARKING_LOT_CAPACITY = 14
+ACCEPTED_TIME_DELAY = 900
+FREE_PARKING_TIME = 900
+HOURLY_TARIFF = 2.5
+DAILY_TARIFF = 10
+SECONDS_IN_HOUR = 3600
+SECONDS_IN_DAY = 86400
+
+
+class TariffName(str, Enum):
+    hourly = "hourly"
+    daily = "daily"
+
+
+class Tariffs(int, Enum):
+    hourly = HOURLY_TARIFF
+    daily = DAILY_TARIFF
+
+
+class CarAlreadyRemovedException(Exception):
+    def __init__(self, car_id):
+        self.car_id = car_id
+
+
+class InvalidLocationException(Exception):
+    def __init__(self, location):
+        self.location = location
+
+
+class LocationNotAvailableException(Exception):
+    def __init__(self, location):
+        self.location = location
+
+
+class StartDateException(Exception):
+    def __init__(self, start):
+        self.start = start
+
+
+class CarIdExistsException(Exception):
+    def __init__(self, car_id):
+        self.car_id = car_id
+
+
+class CarIdDoesNotExistException(Exception):
+    def __init__(self, car_id):
+        self.car_id = car_id
+
+
+class ParkedCar(BaseModel):
+    car_id: str = Field(
+        default=None, title="The licence plate of the car", max_length=8
+    )
+    status: str = Field(default=None)
+    tariff: TariffName | None
+    location: int = Field(ge=0, lt=PARKING_LOT_CAPACITY, description=f"The location must be between 0 and {PARKING_LOT_CAPACITY}")
+    start: datetime | None
+    end: datetime | None = None
+    fee: float | None = Field(default=None, ge=0, description=f"The fee must be greater or equal to 0.")
+
+    @validator('location')
+    def location_must_be_greater_than_zero(cls, v) -> int:
+        if v < 0:
+            raise ValueError('must be greater than zero')
+        return v
+
+    def get_exit_fee(self, start: datetime, end: datetime, tariff: float) -> float:
+        start_date = start.replace(tzinfo=None)
+        time_spent_when_parked = (end - start_date).total_seconds()
+
+        if time_spent_when_parked < FREE_PARKING_TIME:
+            return 0
+        else:
+            match tariff:
+                case TariffName.hourly:
+                    exit_fee = math.ceil(time_spent_when_parked / SECONDS_IN_HOUR) * Tariffs.hourly
+                    return exit_fee
+                case TariffName.daily:
+                    exit_fee = math.ceil(time_spent_when_parked / SECONDS_IN_DAY) * Tariffs.daily
+                    return exit_fee
```

Comment:
What would be required to dynamically update tariffs and their algorithm ?

Reply:
we can have a set of predefined algorithms and by using some configuration endpoints update tariffs and chose between the already predifined algorithms